Acqmon Deadtime Plugin

AlexElykov opened the pull request:
**What is the problem / what does the code in this PR do**
Calculating deadtimes for busy/veto monitor such that we can use them later for NODIAQ busy monitoring

**Can you briefly describe how it works?**
Looks for start and end times of veto signals and finds the intervals in between them

**Can you give a minimal working example (or illustrate with a figure)?**
Seems to find the correct deadtime intervals.
![sample_pic](https://user-images.githubusercontent.com/28751746/92136136-f626a480-ee0b-11ea-8ed0-534b561f335e.png)


**Additional info:** I have another plugin that uses the new "veto_intervals" to calculate cumulative deadtimes and put it into a collection on mongo. However, we need to decide in which collection it should go and which fields we want. @darrylmasson 

**Test notebook can be found here**: https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenon1t:alexelykov:am_veto_interval_notebook

AlexElykov commented:
CodeFator issue just fixed by adding few more lines at the end the acqmon_processing here: c1b9731

darrylmasson commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+            for t in ['time', 'endtime', 'interval']:
+                res[v + t] = result[v + 'veto'][t]       
+        
+        return res
```
So each line in the return dtype has the generic start and stop times as required by strax, and then also start and stop for each of the 3 veto types (normal busy, HE busy, and HEV). What happens if one chunk has 6 normal busies, 1 HE busy, and 2 HEV? Would the better method be to just have each line be just start, stop, and veto type?

AlexElykov replied:
I don't remember the exact thing, but as far as I can recall strax didn't like it when I've tried to implement something like that.  What would be the issue with having multiple veto types in one chunk? If the obtained veto interval is !=0 it would populate the resulting dict with an appropriate dtype label. The fake start/end times would be simply derived from the start time of the interval.

darrylmasson replied:
My question was, what does the output of this plugin look like with this dtype? Can you provide an example?

AlexElykov replied:
I've just upladed a short the test notebook to our wiki:
https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenon1t:alexelykov:am_veto_interval_notebook

Hope this helps to illustrate things a bit more clearly.

darrylmasson replied:
This is the cell I was looking for:
```
array([(1595522722021768350, 1595522722021773350, 1595522725885930240, 1595522726048534880, 162604640, 0, 0, 0, 0, 0, 0),
       (1595522722021773350, 1595522722021778350, 1595522726828730740, 1595522726928319730,  99588990, 0, 0, 0, 0, 0, 0),
       (1595522722021778350, 1595522722021783350, 1595522729658690940, 1595522729677668950,  18978010, 0, 0, 0, 0, 0, 0),
       (1595522722021783350, 1595522722021788350, 1595522733431900310, 1595522733514186030,  82285720, 0, 0, 0, 0, 0, 0),
       (1595522722021788350, 1595522722021793350, 1595522736262083060, 1595522736344731590,  82648530, 0, 0, 0, 0, 0, 0)],
      dtype=[(('faux time [ns]', 'time'), '<i8'), (('faux endtime [ns]', 'endtime'), '<i8'), (('Start busy_veto time since unix epoch [ns]', 'busy_time'), '<i8'), (('Stop busy_veto time since unix epoch [ns]', 'busy_endtime'), '<i8'), (('Duration of busy_veto time since unix epoch [ns]', 'busy_interval'), '<i8'), (('Start hev_veto time since unix epoch [ns]', 'hev_time'), '<i8'), (('Stop hev_veto time since unix epoch [ns]', 'hev_endtime'), '<i8'), (('Duration of hev_veto time since unix epoch [ns]', 'hev_interval'), '<i8'), (('Start he_veto time since unix epoch [ns]', 'he_time'), '<i8'), (('Stop he_veto time since unix epoch [ns]', 'he_endtime'), '<i8'), (('Duration of he_veto time since unix epoch [ns]', 'he_interval'), '<i8')])
```
This is storing 11 values per busy/veto interval with >50% zero-padding and unused fields, when we could rather just deal with 3: start, stop, and type. To calculate the total deadtime for a run you would need to do (stop-start) for three separate columns, and this feels really clunky to me.

AlexElykov replied:
Alright, if I have some spare time I'll look into reducing the dtype to the more essential parts, to steamline this a bit.

AlexElykov replied:
@darrylmasson 
I redid the veto data kind, to a more streamlined version. Indeed looks better. Is that what you had in mind?

> array([(1595522725885930240, 1595522726048534880, 162604640, 'busy_veto'),
       (1595522726828730740, 1595522726928319730,  99588990, 'busy_veto'),
       (1595522729658690940, 1595522729677668950,  18978010, 'busy_veto'),
       (1595522733431900310, 1595522733514186030,  82285720, 'busy_veto'),
       (1595522736262083060, 1595522736344731590,  82648530, 'busy_veto')],
      dtype=[(('veto start time [ns]', 'time'), '<i8'), (('veto end time [ns]', 'endtime'), '<i8'), (('veto interval [ns]', 'veto_interval'), '<i8'), (('veto signal type', 'veto_type'), '<U10')])

darrylmasson replied:
Pretty much, yeah. I thought `veto_type` could be a single number (0=busy, 1=HE busy, 2=HEV), but a string is probably more end-user-friendly.

darrylmasson reviewed (changes requested):
I don't know if the dtype works

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+
+    aqmon_channel_names = ('sum_wf','m_veto_sync',
+                           'hev_stop', 'hev_start', 
+                           'bb_stop','bb_start',
```
I guess I can change it to "he", just felt that it would be too confusing to have "hev" and "he"

darrylmasson replied:
I mean, "high energy veto" and "high energy channels" are inherently somewhat confusing, but that's what we decided on, and I think we should at least try to be consistent.

AlexElykov replied:
Sure no problem.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+    """
+        
+    __version__ ='0.0.19'
+    depends_on = ('raw_records_aqmon') 
```
Where, why it would complain? I ran it both in a notebook and also as a integrated plugin in my straxen version, didn't see any error messages.

darrylmasson replied:
https://github.com/AxFoundation/strax/blob/master/strax/plugin.py#L53, also everything I could find assumed it was a tuple.

AlexElykov replied:
You're right but this cool function takes care of it for me :)
https://github.com/AxFoundation/strax/blob/master/strax/plugin.py#L104

AlexElykov replied:
One can use smthing lile:  depends_on = ('tup'),  depends_on= 'tup' and it would be converted to  ('tup',)

darrylmasson replied:
Ah, true, I had forgotten that line

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+        self.channel_range = self.config['channel_map']['aqmon']
+        self.channel_numbers = np.arange(self.channel_range[1]-7,self.channel_range[1] + 1, 1)
+        self.channel_map = dict(zip(self.aqmon_channel_names,self.channel_numbers))
+        self.t0 = self.config['run_start_time']
```
This might be needed depending on what we want to do with inserting these data to mongo. If we save them here or in another plugin to a dedicated collection we need the run_id imho.

darrylmasson replied:
Neither of these quantities are the run id, though.

AlexElykov replied:
Sorry, got confused for a sec...doing smth else in parallel. I've explained the rest above, the run start time is a relic from the times I wanted to get dynamic deadtime calculation in this plugin, I can remove it.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+                 default = int(0), help='time of start run (s since unix epoch)')
+)
+
+@export
```
Hmmm...seems to work but I'll move it anyways.

darrylmasson replied:
I don't know enough to say confidently, but having an order that differs from the usual usage might cause some obscure errors?

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+import straxen
+import numba
+import numpy as np
+from numba import njit
```
True

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+def _mask(x, mask):
+    return x[mask]
+
+@export
```
I think so.

darrylmasson replied:
Where else is it used?

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+    
+    def setup(self):
+        self.channel_range = self.config['channel_map']['aqmon']
+        self.channel_numbers = np.arange(self.channel_range[1]-7,self.channel_range[1] + 1, 1)
```
You're right about the first one. Ch range and ch numbers are kept as atm Ricardo is working on a code/plugin for the m_veto sync pulses, some of which might end up in this plugin. So I wanted to have more flexibility for channel selection in the plugin itself if needed.

WenzDaniel commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+        dtype = [(('faux time [ns]', 'time'), np.int64)]
+        dtype += [(('faux endtime [ns]', 'endtime'), np.int64)]
+        for veto in ('busy_','hev_','bb_'):
+            dtype += [(('Start '+ veto +'veto time since unix epoch [ns]', veto + 'time'), np.int64)]
```
I would say yes. Does not hurt at least.

WenzDaniel commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+            
+@numba.njit
+def _mask(x, mask):
+    return x[mask]
```
Can you provide an MWE about the performance? Is it really so much faster than without numba?

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+    strax.Option('channel_map', track = False, type = immutabledict, 
+                 help = 'frozendict mapping subdetector to (min, max)'
+                      'channel number.'),
+    strax.Option('run_start_time', type = float, track = False, 
```
This is a relic from an older version, as I've already replied to Darryl, will remove it.
Also how the start time is available via chunks?

WenzDaniel replied:
`def compute(*depends_on, start, end): `

AlexElykov replied:
> `def compute(*depends_on, start, end): `

Cheers, seen that before but forgot.
Hope the night shift is going ok LOL

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+    tmp_dtype = strax.time_fields + [(('Interval between start and stop in [ns]', 'interval'), np.int64)]
+
+    def infer_dtype(self):
+        dtype = [(('faux time [ns]', 'time'), np.int64)]
```
It is a fake time to keep strax happy, as imho plugins require "time" and "endtime" as part of their data type to work and its based on the start of the veto interval. Darryl suggested to streamline the final data type, if I'll have time I'll do it. Otherwise for now it would have to stay like that.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+        rec = strax.raw_to_records(raw_records_aqmon)
+        strax.zero_out_of_bounds(rec)
+        strax.baseline(rec, baseline_samples=10, flip=True)
+        hits = strax.find_hits(rec)
```
What do you mean, threshold for the hits? These are lone NIM pulses anything high enough above the baseline should work. However, depending on what Ricardo had in mind for his m_veto sync plugin/code we might put the hit finding for the AM somewhere else, and set some threshold there.

WenzDaniel replied:
I think you should just make sure that you do not find channel noise as a NIM signal. Does not hurt to put there a threshold of idk 50 adc counts.

AlexElykov replied:
> I think you should just make sure that you do not find channel noise as a NIM signal. Does not hurt to put there a threshold of idk 50 adc counts.

I was thinking about having a different small plugin that finds hits in all channels, which can be then used for calculating all the different things like vetos, GPS signals, sum_wf ... can set the thr. there.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+@numba.njit
+def channel_select(rr, ch_stop, ch_start):
+    """Return """
+    return _mask(rr, (rr['channel'] >= ch_stop) & (rr['channel'] <= ch_start)) 
```
1. If I got the question correctly, the answer is the physical inputs on the AM board itself that go as stop -> start, stop -> start, stop->start for all the vetos i.e. 806  = busy_stop and 807 = busy_start.
2. I think so, as I want to select only 2 channels each time.

WenzDaniel replied:
Okay, about 1.) could you put some comment or extend the doc string. I mean the function is very generic in what it is doing and most people wont know, that the first channel is the stop signal and the second one is the start signal.

AlexElykov replied:
> Okay, about 1.) could you put some comment or extend the doc string. I mean the function is very generic in what it is doing and most people wont know, that the first channel is the stop signal and the se

Kk. Will add a comment and some more description of the AM channels.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+            result_temp = np.zeros(len(start),dtype=dtype)
+            result_temp['time'] = start 
+            result_temp['endtime'] = stop
+            result_temp['interval'] = stop - start
```
The interval dtype was meant to be used for live monitoring, so I think it's fine having it. Although if in the future I change the dtype from this plugin only to start_time and end_time of each veto, this calculation can go somewhere else.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,131 @@
+            result = result_temp[result_temp['interval'][:] > 10]
+        else:
+            result = np.zeros(1,dtype=dtype)
+            result['time'] =  t
```
Don't think so, as 'time' used here is converted to *_time (* = hev, busy, bb) above. So imho in principle it can be 0.

WenzDaniel replied:
Okay, maybe just leave a comment then. Right now it is a bit confusing.

AlexElykov commented:
**I've made some modifications to the acqmon plugin.**

- Added a small plugin (producing acqmon_hits) which should allow more flexibility for looking into non veto channels of the AM such as the GPS sync and maybe sum_wf
- Addressed the previously mentioned issues
- Changed the veto_intervals data kind to a more streamlined version based on Darryl's suggestion

jorana commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,138 @@
+    __version__ ='0.0.2'
+    
+    depends_on = ('raw_records_aqmon') 
+    provides  = ('acqmon_hits')
```
This I find somewhat confusing. I'd prefer to also call this aqmon rather than acqmon otherwise it's inconsistent with the raw_records_aqmon

AlexElykov replied:
It's acutally "a**c**quisition" not "aquisition" but I can chage it to aqmon if you like it better

jorana reviewed (commented):
Hi Alex. Hereby some requests for changes. 

I've looked at your PR and it generally looks good. Also the tests and output look intuitive https://github.com/XENONnT/analysiscode/blob/master/StraxTests/AqMon.ipynb.

I'll fix the failing travis in a bit. That is actually my fault rather than yours.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,138 @@
+    depends_on = ('raw_records_aqmon') 
+    provides  = ('acqmon_hits')
+    data_kind = ('acqmon_hits')
+    
```
Ok will change it to TARGET

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,138 @@
+    def compute(self, raw_records_aqmon):
+        rec = strax.raw_to_records(raw_records_aqmon)
+        strax.zero_out_of_bounds(rec)
+        strax.baseline(rec, baseline_samples = 10, flip = True)
```
Yeap, will do.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,138 @@
+    """Return data in between stop and start channels of the acquisition monitor (AM). 
+       The order is reversed becasue physical inputs on the AM board go as stop -> start
+       for all the veto signals i.e. 806 = busy_stop and 807 = busy_start """
+    return _mask(rr, (rr['channel'] >= ch_stop) & (rr['channel'] <= ch_start)) 
```
Daniel made a test somewhere above, it indeed seems to be faster.

jorana replied:
Sure, that is fine but the `>=` is not what you want, you want exactly ch_stop and ch_start

AlexElykov replied:
Yeap, makes sense. Changed.

AlexElykov commented on straxen/contexts.py:
```diff
@@ -11,7 +11,8 @@
         straxen.nveto_pulse_processing,
-        straxen.nveto_hitlets],
+        straxen.nveto_hitlets,
+	straxen.acqmon_processing],
```
Will fix.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,138 @@
+    tmp_dtype = strax.time_fields
+
+    def infer_dtype(self):
+        dtype = [(('veto start time since unix epoch [ns]', 'time'), np.int64)]
```
Ok, done.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,138 @@
+    data_kind = ('veto_intervals')
+
+
+    aqmon_channel_names = ('sum_wf','m_veto_sync',
```
Moved

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,138 @@
+    
+    def setup(self):
+        # Keeping ch_numbers and ch_range as class attributes for now, maybe will need them for other features
+        self.channel_numbers = np.arange(800,808, 1)
```
Cheers, will do.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,138 @@
+                           'he_stop','he_start',
+                           'busy_stop', 'busy_start')
+    
+    tmp_dtype = strax.time_fields
```
Moved.

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,138 @@
+def _mask(x, mask):
+    return x[mask]
+
+@export
```
Alright, removed it.

AlexElykov commented:
Hi, @WenzDaniel can you please review the recent updates and approve the PR. I think I've addressed all the requests/concerns that were raised thus far.

WenzDaniel commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,132 @@
+                                           dtype = self.tmp_dtype, t = 0)
+            
+            # Check that we found a veto start time and then populate the resulting dict
+            if np.any(vetos['time']):
```
Minor comment, you can use `if len(vetos):` here should give you the same result and is roughly x40 faster since you do not have to check the content of every entry.

Okay took me a while... but now I also understand why merge_vetos has 

```
else:
    result = np.zeros(1,dtype=dtype)
    result['time'] =  t
    result['endtime'] = t
```

with t=0. I think a slightly neater way to do it is to change it in the following way:

 ```
else:
    result = np.zeros(0, dtype=dtype)
```

and then later the suggested `if len(vetos):` 

Otherwise it looks like you create a forbidden zero length interval, but actually this is not true since you call `np.any(vetos)`. However I think in this way it is super confusing.

AlexElykov replied:
Ok, seems reasonbale to me, will change.

WenzDaniel commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,132 @@
+            if np.any(vetos['time']):
+                res['time'] = vetos['time']
+                res['endtime'] = vetos['endtime']
+                res['veto_interval'] = vetos['endtime'] - vetos['time']
```
Is not it redundant? You could save 1/3 of the data_type size by removing this field. Computing it should be fairly fast and simple too.

AlexElykov replied:
I like it, as if we plug the result of this plugin straight into DB it would give me fewer headaches of calculating things in NODIAQ. This data kind was already tiny and I've made it only a fraction of the original one, so data size is also not an issue.

WenzDaniel replied:
Okay, I am fine with it

WenzDaniel commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,132 @@
+def channel_select(rr, ch_stop, ch_start):
+    """Return data in between stop and start channels of the acquisition monitor (AM)"""
+    return rr[((rr['channel'] == ch_stop) | (rr['channel'] == ch_start))] 
+
```
Code factor is complaining here, might it be that this last line is not empty?

AlexElykov replied:
Will fix

WenzDaniel commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,132 @@
+        hits = aqmon_hits
+        
+        res = dict()
+        for i, veto in enumerate(self.veto_names):
```
Maybe it is still too early in the morning, but I think your loop is not want you want is it?

I think you are always overwriting your results in the dictionary `res` are not you?

MWE:
```
res = dict()

dtype = [(('bla', 'a'), np.int64), (('bla 2', 'b'), np.int64)]
for i in range(1,3):
    print(i)
    sub_res = np.ones(i, dtype) 
    sub_res['a'] = sub_res['a'] * i
    res['a'] = sub_res['a']
    res['b'] = sub_res['b']
res
```

which will give you res['a'] = (2, 2) instead of (1, 2, 2).

AlexElykov replied:
Cheers! Good catch, it just teaches me not to code under the effect of painkillers. Will fix.

WenzDaniel commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,132 @@
+            result['time'] = start 
+            result['endtime'] = stop
+        else:
+            result = np.zeros(1,dtype=dtype)
```
See comment above

WenzDaniel reviewed (changes requested):
Hey Alex, I am sorry, but in case I am not still half asleep due to the LXe filling shifts, I think your code is not working as you would expect. Could you please address my comments and ensure with your example notebook that it is really working as intended?

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,132 @@
+        return dtype
+    
+    def setup(self):
+        aq_ch_start = 800
```
This was suggested here by Joran: https://github.com/XENONnT/straxen/pull/207#discussion_r486989546
I highly doubt that we are going to change the channel map or hardware cabling at this point, and if we do any such change will still mean that we need to change all the mapping. 

Nonetheless, my suggestion would be to split the aqmon channels in the straxen context, https://github.com/AlexElykov/straxen/blob/56abca546b2e6f31ff3a49b04513cd6ba3fe00bb/straxen/contexts.py#L31
to something like this: 

soft_aqmon = (790,799)    
hard_aqmon = (800, 807)

Then I can use the hard_aqmon and it would be easy to track the changes.

WenzDaniel replied:
Okay I see, but maybe we can then find a better place for this hard coded number? I think so far we put all hard coded value which we should keep track of in the beaning of  common.py. Would this be okay with you?

darrylmasson replied:
Daniel I'm not sure if you do see Alex's point. This plugin needs the hardware AM channels, but the global channel map doesn't distinguish between the hardware and software channels, it just lumps everything together.

WenzDaniel replied:
No No thanks, this I understood from Alex comment. And I agree on his point, that we should not split up the aqmon channels on the raw_records level. I just would like to have a good bookkeeping of all really important hard coded numbers. I think this is the case here since in principle I can imagine that also some other plugin may distinguish between the "soft" and "hard" aqmon channels. And hence having both plugins, functions or whatever referring to the same parameter in common.py will give us less headaches in case we change something (even though this might be unlikely, but you never know).

darrylmasson replied:
As far as I know there's nothing past DAQreader that uses the software channels, and this is where all the software-based deadtime signals show up from both strax and redax. Currently these are ignored. I agree that we don't want separate files for the SW and HW components, but it'll be really handy to have this distinction hard-coded somewhere sensible. If this sensible location is common.py or in the context, I don't know.

AlexElykov replied:
Alright, then I'll just add it to common.py as a variable. Any requests for its name. Something like "n_hard_aqmon_start = 800" would be fine with you @WenzDaniel ?

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,132 @@
+        
+        res = dict()
+        for i, veto in enumerate(self.veto_names):
+            channels = channel_select(hits, self.channel_map[veto + 'stop'], self.channel_map[veto + 'start'])
```
It is the function that does select the channels, name seems appropriate to me.

WenzDaniel replied:
Yes, but the thing which you return are not channels, are they? They are some sort of "hits" while to me channels implies something like an array of intigers which represent channel numbers.

AlexElykov replied:
Ok, how about "veto_hits" then for its name?

AlexElykov commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,132 @@
+                               'hev_stop', 'hev_start', 'he_stop','he_start','busy_stop', 'busy_start')
+        channel_numbers = np.arange(aq_ch_start, aq_ch_start + len(aqmon_channel_names))
+        
+        self.tmp_dtype = strax.time_fields
```
Ok. I'll put it into the function.

AlexElykov commented:
@jorana @WenzDaniel 
Made additional fixes to the aqmon based on your input:

- Added the deault start channel for hardware aqmon in common.py
- Fixed a bug in the resulting dict when enries were overwritten by new ones.

Tested with few runs, veto intervals seem to be correct. Simple test notebok can be found here:
https://gist.github.com/AlexElykov/b804b03f1951dfe534a56c3b51487059

jorana commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,130 @@
+            vetos = self.merge_vetos(veto_hits, gap = self.config['max_veto_gap'],\
+                                           dtype = strax.time_fields, t = 0)
+            
+            # Check that we found a veto interval and update the resulting dict
```
Not a major comment but usually we'd make a large buffer of the dtype you need and then clip until the length you actually added things. Might be somewhat faster.

Would look something like:
```python
buff_len = 10_000
res = np.zeros(buff_len, dtype = self.dtype)
n_res = 0

if len(vetos):
   res[n_res]['time'] = vetos['time']
   ... <same for others>
   n_res += 1
   if n_res > buff_len:
      raise ValueError('ran out of buffer') 
res[:n_res']
```

AlexElykov replied:
Cheers, though, I think I've seen straxen plugins implementations without buffering first. Also, these data are so tiny compared to anything else that we will process, so I don't think it will make a difference. If you think that this is a show stopper, let me know, and I'd redo it again. Otherwise, I would keep at as it is..

jorana replied:
sure

WenzDaniel commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,130 @@
+        self.channel_range = self.config['channel_map']['aqmon']
+        self.channel_map = dict(zip(aqmon_channel_names, channel_numbers))
+        self.veto_names = ['busy_','he_','hev_']
+        return self.channel_map
```
@jorana is the setup return used by anything in strax? I do not think so.

jorana replied:
Sharp. I don't recall any instance of such a usage of return

AlexElykov replied:
I think I've seen it somewhere but can't remember where. I can remove it, though it seems to be harmless, or I'm missing something?

WenzDaniel replied:
Yes, you can just remove it

AlexElykov replied:
Why does it matter though?

WenzDaniel commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,130 @@
+                res.setdefault("endtime",[]).extend(vetos['endtime'])
+                res.setdefault("veto_interval",[]).extend((vetos['endtime'] - vetos['time']))
+                res.setdefault("veto_type",[]).extend([veto + 'veto']*len(vetos))
+        return res
```
I might be wrong, but in case of more than one veto_name in `veto_names`  the result is not sorted by `time` anymore. Could you fix this? I think it is not mandatory by strax, but all our other data_kinds are sorted by time.

jorana replied:
They are sorted by time for a given channel, so in principle this isn't that different. See:
https://github.com/AxFoundation/strax/blob/master/strax/processing/general.py#L25

Although it doesn't harm to do it here too

AlexElykov replied:
Hmm...I think they will be sorted though. Physically there can be only a single veto each time, either normal busy, he or hev. So they are bound to come one after another, but maybe I misunderstood your comment.

WenzDaniel replied:
The issue is, the way you are calling your functions in your for loop you will end up with the following result:

```
res = [veto_busy_1, veto_busy_2, veto_busy_3..... veto_busy_n, veto_he_1, veto_he_2, .... veto_he_n, veto_nv_1, veto_nv_2 ...]
```

(symbolically speaking ). Hence your result will be sorted first by channel and then by time, while for all other data_types we sort everything by time and then by channel.

AlexElykov replied:
Ok, but as far as I know dicts are by their nature unsorted. So any returned dict from a plugin will be unsorted i.e. your comment would apply to stuff like:
https://github.com/XENONnT/straxen/blob/ed27f9c105bcdd9144d754d8d6f31d5841e1c40e/straxen/plugins/peak_processing.py#L257

WenzDaniel replied:
> Ok, but as far as I know dicts are by their nature unsorted. 

You are mixing things up here. Yes the keys with respect to each other are unsorted, but not the values within a key. E.g. 

```
d = dict(a=[1,2,3], v=2)
print(d['a'])
```
will always yield [1,2,3] not [2,1,3] nor [3,2,1].

darrylmasson replied:
For the record, from the python documentation:
> Changed in version 3.7: Dictionary order is guaranteed to be insertion order. This behavior was an implementation detail of CPython from 3.6.

AlexElykov replied:
> > Ok, but as far as I know dicts are by their nature unsorted.
> 
> You are mixing things up here. Yes the keys with respect to each other are unsorted, but not the values within a key. E.g.
> 
> ```
> d = dict(a=[1,2,3], v=2)
> print(d['a'])
> ```
> 
> will always yield [1,2,3] not [2,1,3] nor [3,2,1].

Ok, now I understand what you mean. The question I have though, is what does this lack of sorting affects?

WenzDaniel replied:
Currently I think it just confuses the user. But I cannot exclude that their might at some point a check testing if the data is sorted. So it is simply better to be consistent and it is not a big modification.

AlexElykov replied:
Alright. Then I'll just add:

res_temp = strax.dict_to_rec(res)
res_temp.sort(order = 'time')

To sort out and output the result by time.

WenzDaniel reviewed (approved):
Only two minor comments. The rest looks fine. Thanks a lot.

AlexElykov commented:
Alright, more updates based on the input, with the 2 main ones being:

- Removed the return from the setup function
- Result from plugin is sorted by "time"

Plugin was tested with the same notebook as before, seems to work.

WenzDaniel commented on straxen/plugins/acqmon_processing.py:
```diff
@@ -0,0 +1,133 @@
+                res_temp.setdefault("veto_interval",[]).extend((vetos['endtime'] - vetos['time']))
+                res_temp.setdefault("veto_type",[]).extend([veto + 'veto']*len(vetos))
+        
+        res = strax.dict_to_rec(res_temp)
```
There seems to be an error with empty dicts: 

```
x = {}, dtype = None

    @export

    def dict_to_rec(x, dtype=None):

        """Convert dictionary {field_name: array} to record array

        Optionally, provide dtype

        """

        if isinstance(x, np.ndarray):

            return x

    

        if dtype is None:

            if not len(x):

>               raise ValueError("Cannot infer dtype from empty dict")

E               ValueError: Cannot infer dtype from empty dict
```

jorana replied:
```python
res = strax.dict_to_rec(res_temp, dtype = self.dtype)
```

WenzDaniel replied:
Yes, sorry forgot to copy the line

jorana replied:
OK, yeah it's the trivial solution to fix Travis. Alex I just added it in d92de8c. I believe the review has been quite thorough so let's merge it! Would love to have this shown on the website.

AlexElykov replied:
@WenzDaniel @jorana Cheers guys :)
